lib/plug/cowboy/drainer.ex

```diff
@@ -0,0 +1,79 @@
+defmodule Plug.Cowboy.Drainer do
+  @moduledoc """
+  TODO: description
+  """
+  use GenServer
+
+  @doc """
+  TODO: description
+
+  ## Options
+
+    * `:id` - The ID for the process.
+      Defaults to `Plug.Cowboy.Drainer`.
+
+    * `:shutdown` - How long to wait for connections to drain.
+      Defaults to 5000ms.
+
+    * `:refs` - The cowboy listener refs to drain connections for.
+      Defaults to all cowboy listeners.
+
+    * `:drain_check_interval` - How frequently to check if a listener's
+      connections have been drained.
+      Defaults to 1000ms.
+  """
+  @spec child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
+  def child_spec(opts) when is_list(opts) do
+    {spec_opts, opts} = Keyword.split(opts, [:id, :shutdown])
+
+    Supervisor.child_spec(
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, [opts]},
+        type: :worker
+      },
+      spec_opts
+    )
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  @impl true
+  def init(opts) do
+    Process.flag(:trap_exit, true)
+    {:ok, opts}
+  end
+
+  @doc false
+  @impl true
+  def terminate(_reason, opts) do
+    opts
+    |> Keyword.get(:refs)
+    |> drain(Keyword.get(opts, :drain_check_interval, 1_000))
+  end
+
+  defp drain(nil, drain_check_interval) do
+    :ranch.info()
+    |> Enum.map(&elem(&1, 0))
+    |> drain(drain_check_interval)
+  end
+
+  defp drain(refs, drain_check_interval) do
+    refs
+    |> Stream.each(&suspend/1)
```

Comment:
We should run each of those in a separate `Task.start_link` so we can actually shut down concurrently.

Reply:
I had started with using `Task.async_stream/2`. However, once the listener has been suspended, everything is happening concurrently already. The waiting here is technically happening in sequence, but the thing being waited on is happening concurrently. So, if one listener took 2 seconds to drain, and the next one took 1 second, once it gets past waiting for the first, the second will already be drained and not wait.

I can add the task stuff back as a PR to this branch for comparison.

Reply:
Ah, good call. So this is good to ship it. Just one tiny change then:

```suggestion
    |> Enum.each(&suspend/1)
```

A stream  is most likely not buying us anything here.

Reply:
The return of `Enum.each/2` is `:ok`. The return of `Stream.each/2` is the initial collection.

Reply:
BTW, I would love to if that changed 😉

Reply:
Oh, of course. We made it return `:ok` because people would write imperative code inside each and wonder what happened.